source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * This is a specialized structure intended for static header maps, but
+ * there may be other use cases.
+ * The structure is:
+ * 1. a length-based lookup table so only keys the same length as the
+ * target key are considered.
+ * 2. a trie that branches on the "most divisions" position of the key.
+ *
+ * For example, if we consider the case where the set of headers is
+ * `x-prefix-banana`
+ * `x-prefix-babana`
+ * `x-prefix-apple`
+ * `x-prefix-pineapple`
+ * `x-prefix-barana`
+ * `x-prefix-banaka`
+ *
+ * A standard front-first trie looking for `x-prefix-banana` would walk
+ * 7 nodes through the tree, first for `x`, then for `-`, etc.
+ *
+ * This structure first jumps to matching length, eliminating in this
+ * example case apple and pineapple.
+ * Then the "best split" node is on
+ *   `x-prefix-banana`
+ *               ^
+ * so the first node has 3 non-miss branches, n, b and r for that position.
+ * Down that n branch, the "best split" is on
+ *   `x-prefix-banana`
+ *                 ^
+ * which has two branches, n or k.
+ * Down the n branch is the leaf node (only `x-prefix-banana` remains) - at
+ * this point a regular string-compare checks if the key is an exact match
+ * for the string node.
+ */
+template <class Value> class CompiledStringMap {
+  using FindFn = std::function<Value(const absl::string_view&)>;
+
+public:
+  using KV = std::pair<std::string, Value>;
+  /**
+   * Returns the value with a matching key, or the default value
+   * (typically nullptr) if the key was not present.
+   * @param key the key to look up.
+   */
+  Value find(const absl::string_view& key) const {
+    if (key.size() >= table_.size() || table_[key.size()] == nullptr) {
```

Comment:
I tried taking it to a local variable, but that has to be after key.size() being too large has already been checked, which makes it two if-branches, which from benchmarking appears to make both the hit paths ~5% slower.

It's not like these are actual table *lookups*, they're just offsets, so even if it wasn't optimized away it might still be faster than taking an additional copy.

Reply:
`table_[key.size()]` is an out-of-bounds crash if key.size() > table_.size().

Reply:
```
const size_t key_size = key.size();
if (key_size >= table_.size() || table_[key_size] == nullptr) {
...
```